twine/commands/check.py

```diff
@@ -68,47 +69,74 @@ def __str__(self):
         return self.output.getvalue()
 
 
+def _description_rendering_errors(description_content_type, description):
+    warning_stream = _WarningStream()
+    content_type, params = cgi.parse_header(description_content_type)
+    renderer = _RENDERERS.get(content_type, _RENDERERS[None])
+    rendering_has_errors = (
+        description is not None
+        and renderer
+        and renderer.render(description, stream=warning_stream, **params)
+        is None
+    )
+    errors = []
+    if rendering_has_errors:
+        errors.append(
+            "The project's long_description has invalid markup which "
+            "will not be rendered on PyPI. The following syntax "
+            "errors were detected:\n%s" % warning_stream
+        )
+    return errors
+
+
+def _description_content_type_errors(description_content_type):
+    description_content_type_invalid = (
+            description_content_type not in _VALID_CONTENT_TYPES
+    )
+    errors = []
+    if description_content_type_invalid:
+        errors.append(
+            'warning: `long_description_content_type` invalid.\n'
+            'It must be one of the following types: [{}].\n'
+            .format(", ".join(_VALID_CONTENT_TYPES))
+        )
+    return errors
+
+
+def check_package(package, output_stream):
+    metadata = package.metadata_dictionary()
+    description_content_type = metadata["description_content_type"]
+    if description_content_type is None:
+        output_stream.write(
+            'warning: `long_description_content_type` missing.  '
+            'defaulting to `text/x-rst`.\n'
+        )
+        description_content_type = 'text/x-rst'
+    description = metadata["description"]
+    if description in {None, 'UNKNOWN\n\n\n'}:
+        output_stream.write('warning: `long_description` missing.\n')
+        description = None
+    failures = (
```

Comment:
I realize the future of this PR is in question, but I'm curious about the decision to inline some validation, and put others in helper functions. FWIW, I'd find this easier to follow if they were all in `check_package`, and given the uncertainty around what should be a warning or a failure, I think it'd be easier to change in the future.

Reply:
My thinking here is that warning being emitted is not relevant to the final result of `check_package`. It does not feel like a validation so much as a log line since a package cannot fail due to this